### ISSUE TITLE
UAA support: should it be removed ?

### ISSUE THREAD

**Issue by pascalgrimaud** [jhipster/generator-jhipster#13081]
@xetys : you did a great job for UAA support, but today, it's not well maintained.
See https://github.com/hipster-labs/jhipster-daily-builds/actions?query=workflow%3A%22Microservices+UAA%22

As the support of UAA is mainly done by 1 person, I open this ticket to discuss if someone else wants to maintain this option ?
If not, the option could be simply removed, as it is useless to keep an option which is broken since months.

Then, the option is not maintained in JHipster Registry too, so it could be removed there too.

So plz comment.

**Comment by pascalgrimaud** [comment]
cc @jhipster/developers

**Comment by avdev4j** [comment]
hi,
My 2 cents
I would say UAA should be moved to a dedicated blueprint. By doing this, it will target a specific version of JHipster, the maintenance will be easier and it will avoid to break features (or builds) in the main generator if there is new breaking changes for example.

regards,

**Comment by xetys** [comment]
if moving to a blueprint, would it still be possible to use it in a JDL like today?

I see the issue with maintenance, I'm kinda far from contributing the past year. I'd love to find a way back to this, but I'm with you, other contributors to UAA would be beneficial. I remember there has been a time when a quite notable amount of users were using UAA. Could someone tell me how the current situation is? Would be interesting for me to know if it even makes sense to put afford in this option. I'm still using it every day and it's a great thing

**Comment by MathieuAA** [comment]
Yes you can! As long as options from the blueprints are handled, well, inside the blueprint, everything should be good!

**Comment by pascalgrimaud** [comment]
up, for final v7

**Comment by PierreBesson** [comment]
After 2 months, there has been no further  comments from the community on this. I would say the consensus is to remove UAA support for JHipster v7 as it is currently broken. In the future a blueprint can be created for a next-gen UAA architecture.
Especially with the new blueprint capabilities coming in v7, such a blueprint should be relatively easy to do and will allow the community of UAA users to progress at their own pace without blocking or being blocked by the evolution of the main generator.

**Comment by deepu105** [comment]
Yes lets remove it

Thanks & Regards,
Deepu

**Comment by pascalgrimaud** [comment]
As no one wants to do this, I'm starting it so

**Comment by pascalgrimaud** [comment]
Bounty claimed at https://opencollective.com/generator-jhipster/expenses/33412

**Comment by jfougere** [comment]
What does that mean exactly for users currently having a UAA server & UAA authentication configured ? 
Stuck in V6 ? Need to migrate to Keycloak ?

**Comment by pascalgrimaud** [comment]
@jfougere : as JHipster is a code generator, you can manually upgrade your applications and not rely anymore to JHipster.

You can still try to use v6 but I don't know the exact version when UAA was working.

I'm sorry for removing important options like this one, but it is useless to keep broken options. We were stuck when trying to migrate to Spring Boot 2.3, then 2.4. And after asking help in this ticket, mailing list, twitter, as no one want to maintain this option, the final decision was to simply remove UAA.

Hope you can understand.

**Comment by jfougere** [comment]
Yes I understand, I am just wondering what would be the correct way forward! 
I also try to understand what is now the recommended way to secure application with JHipster 7+ ? 

Regarding the Spring Boot 2.3/2.4 upgrade, what was the pain point ? Was it the deprecated code from `sprint-security-oauth2` ?

**Comment by pascalgrimaud** [comment]
I suggest you to use OAuth2 with Keycloak or Okta :-)

About 2.3, I didn't take care of the migration, so I can't tell.
About 2.4, all gateways are reactive now, using Webflux, because Zuul and Ribbon are deprecated. So it would be a hard part to upgrade.

**Comment by shougao** [comment]
> After 2 months, there has been no further comments from the community on this. I would say the consensus is to remove UAA support for JHipster v7 as it is currently broken. In the future a blueprint can be created for a next-gen UAA architecture.
> Especially with the new blueprint capabilities coming in v7, such a blueprint should be relatively easy to do and will allow the community of UAA users to progress at their own pace without blocking or being blocked by the evolution of the main generator.

wish to know the next-get UAA will release

**Comment by pascalgrimaud** [comment]
@shougao : UAA has been removed from JHipster and I don't think someone started a blueprint, so nothing is planned.

**Comment by kevin-madhu** [comment]
How is the JHipster UAA server superior to the spring authorization server? Is it a viable option to concentrate on bringing https://github.com/spring-projects-experimental/spring-authorization-server to the Jhipster ecosystem? Maybe that'll solve the maintainability issue because it's a spring project itself (still listed as experimental though). Maybe an initiative to bring it to the Jhipster ecosystem could impart more momentum into that project. Any thoughts?

**Comment by jdubois** [comment]
@kevin-madhu they probably share some common code, but JHipster UAA isn't maintained anymore. It was just too much work, and not enough people ready to contribute to it.
For something as important as security, I would recommend a 3rd party service like Okta, Keycloack (but you need to manage it seriously, and take a support contract), Azure Active Directory, etc.

**Comment by kevin-madhu** [comment]
@jdubois Exactly my point. Do you suggest a new option of using "Spring Authorization Server" as a security option in addition to Okta and Keycloack, replacing the JHipster UAA option. Since Spring Authorization Server is an official project of Spring itself, maybe it's more maintained than JHipster UAA.

**Comment by pascalgrimaud** [comment]
@kevin-madhu : it won't be done in main generator-jhipster because it's too much work.
But are you ready to start a blueprint or a module to add this option, and maintain it as your personal project ?

**Comment by kevin-madhu** [comment]
@pascalgrimaud I'm ready to start working on it this weekend.

**Comment by pascalgrimaud** [comment]
@kevin-madhu : good news! don't hesitate to ping `@jhipster` on Twitter, once your project is started, so we can give visibility

**Comment by kevin-madhu** [comment]
@pascalgrimaud Cool! Thanks

**Comment by jfougere** [comment]
I'm no expert in OAuth2 so I may propose a stupid thing, correct me if I'm wrong.
As far as I know the UAA server is a valid OAuth2 server. Why can't we support it as any OAuth2 server using the OAuth2 option ? 

I understand that current OAuth2 option setup things for OpenID. Maybe we could have 2 options: base OAuth2 & OpenID. 
What do u think ?

**Comment by mraible** [comment]
Spring Authorization Server supports OAuth and OIDC, so this should be possible. 

https://spring.io/blog/2021/05/10/spring-authorization-server-0-1-1-available-now

We'll just have to document how to configure it, like we do for Okta.

https://www.jhipster.tech/security/#okta

**Comment by jfougere** [comment]
@mraible  do you think we could keep support for UAA server while using the OAuth2 option in JHipster ? 

I am asking because adding support for Spring Authorization Server is great but in the meantime it would be nice to offer a solution for users still having a UAA server.

**Comment by pascalgrimaud** [comment]
@jfougere : I totally understand your need and we are really sad for removing the UAA support.
The problem is to maintain the option:
- who wants to add back the UAA support ?
- who wants to maintain it, for the next months / years ?
- it means to maintain the option in generator-jhipster, jhipster-bom, jhipster-daily-builds, jhipster-online, jhipster.github.io, jhipster-base, jhipster-registry, etc.

The problem we had with UAA support, is the lack of maintenance. No one wants to maintain it.
We had daily builds which failed during more than 6 months, and when I asked help, no one answered :(

I don't want the option to be added back, and in 6 months, it's broken again.
I don't want to be stuck when we'll upgrade Spring Boot and the UAA option will be broken again.

Another problem is the support is done during our personal time.
That's the main point.

**Comment by jfougere** [comment]
@pascalgrimaud yes I understood perfectly the reason. 
I'm just saying that a UAA server is also an OAuth2 server. So it would be nice to have an option OAuth2 in JHipster (Not OIDC) that allows to connect to whatever OAuth2 server we use (UAA server included).

I suppose the use of OIDC in Spring is not mandatory no ? (I'm no expert again here).

**Comment by pascalgrimaud** [comment]
@jfougere : the expert here is Matt, and as mentioned above, it's possible :-)

In my opinion, how it can be achieved:
- build your Spring Authorization Server (https://spring.io/blog/2021/05/10/spring-authorization-server-0-1-1-available-now)
- generate a project with JHipster + OAuth2
- apply some changes (code / config) to make it works with Spring Authorization Server, as the generated code is yours, you can do what you want -> I don't have any idea how to do this part :)

**Comment by deepu105** [comment]
I also agree about not adding UAA back. Unfortunately we can't maintain
options unless its requested by a lot of people. And as Pascal mentioned it
shouldn't be hard to modify generated code to make it work and if you are
really interested I would suggest building a blueprint that does this

**Comment by jfougere** [comment]
@pascalgrimaud  I understand your point but again it's not exactly what I meant. I'm not asking to put back UAA option neither to maintain the UAA server itself. 

I'm just saying that the UAA is also a valid OAuth2 server (as far as I know), and so supporting OAuth2 should normally mean supporting also the UAA server.

In what way the UAA needs specific client code ? (Here the key point for me is to avoid having to migrate our UAA server to something else at least for now)

**Comment by xetys** [comment]
Slowly I'm waking up from retirement and suddenly get here :D well...

I'm very sorry that I haven't been able to maintain UAA in the last few months, so I understand the decision of removing it from the main generator. For a couple of days, I'm working on a first version of the UAA blueprint. Here I'm proceeding not to fast, as I want to make this blueprint overriding as few as possible from the main generation process. That's quite a bit of work to do for now. 

Anyways, I'm running too many UAA applications to cut them from future upgrades as UAA is not supported, so I'm confident it will return as a blueprint soon. So @kevin-madhu I'm very happy to get some support for the blueprint.


@jfougere about UAA and OAuth2. It's kinda yes and no answer to this. Yes, both OAuth2 (keycloak/okta) are using OAuth2 for the flow. But there are some differences. The OAuth2 option is a little bit more suited for an external IDP. We don't assume the service is being part of your microservices. The UAA on the other hand is an own service, which you can client-load-balance and extend to your own logic. The philosophy here is the following:

- the OAuth2 option allows to fully outsource all user stuff to an external identity provider (IDP). Such as Keycloak or Okta. You are fine handling all user-related stuff outside of your application's scope. Here, OAuth2+OIDC IS the de facto standard how to design communication between your application and the IDP. 
- the UAA option allows you a flexible security setup for the microservice context (user2service calls, service2service calls, 3rd party clients for exposing your API to the outside, etc). Compared to the JWT option, UAA allows you to modify your user design (registration flow, custom user data, e-mail templates, bla bla bla) as you could normally do with JHipster, but you don't have to reinvent the wheel when it comes to common security scenarios, where the JWT option is just not giving you the tools. OAuth2 is used here, because that specification does describe how to handle that.

I'm still the opinion, that UAA gives a great value for JHipster microservice. It's not exactly the same as OAuth2 option.


And while I'm here... I'm planning to set up the UAA blueprint first as it was before it was removed. I am aware that Spring is developing a new version of the authorization server. They wanted to remove it because they thought Keycloak does the job, but as I explained so much in the past: there ARE cases when users want to implement their own OAuth2 authserver and not using a complete external solution. I'm happy to see, that the spring team finally got convinced by the com that this still is a thing and proceed their work on a new service. But as version 0.0.1 doesn't seem stable to me I think to restore what there was, and when things get stable, move to that new stuff then.

**Comment by xetys** [comment]
I don't get what you mean with specific client code. 

If you mean the JHipster client (say Angular), then the client code is simpler than in the JWT case, as the frontend doesn't manage the tokens.

If you mean the gateways code, so UAA provided a very unique thing. The gateway stored the JWT access tokens in a HTTP-only cookie. With this, the frontend didn't had to manage the token anymore. Even more, it is safer this way, as there is no way of performing XSS attacks on tokens as there is no token visible to JS. 

Then again, there is some code for all services to fetch the verification key. In contrast to JWT, UAA uses asymmetric keys. That is indeed a bit specific and possibly can be changed towards a framework standard.

**Comment by jfougere** [comment]
Thanks @xetys, yes I meant code in services. Thanks for the clarification.

**Comment by kevin-madhu** [comment]
> Slowly I'm waking up from retirement and suddenly get here :D well...
> 
> I'm very sorry that I haven't been able to maintain UAA in the last few months, so I understand the decision of removing it from the main generator. For a couple of days, I'm working on a first version of the UAA blueprint. Here I'm proceeding not to fast, as I want to make this blueprint overriding as few as possible from the main generation process. That's quite a bit of work to do for now.
> 
> Anyways, I'm running too many UAA applications to cut them from future upgrades as UAA is not supported, so I'm confident it will return as a blueprint soon. So @kevin-madhu I'm very happy to get some support for the blueprint.
> 
> @jfougere about UAA and OAuth2. It's kinda yes and no answer to this. Yes, both OAuth2 (keycloak/okta) are using OAuth2 for the flow. But there are some differences. The OAuth2 option is a little bit more suited for an external IDP. We don't assume the service is being part of your microservices. The UAA on the other hand is an own service, which you can client-load-balance and extend to your own logic. The philosophy here is the following:
> 
> * the OAuth2 option allows to fully outsource all user stuff to an external identity provider (IDP). Such as Keycloak or Okta. You are fine handling all user-related stuff outside of your application's scope. Here, OAuth2+OIDC IS the de facto standard how to design communication between your application and the IDP.
> * the UAA option allows you a flexible security setup for the microservice context (user2service calls, service2service calls, 3rd party clients for exposing your API to the outside, etc). Compared to the JWT option, UAA also allows you to modify your user design (registration flow, custom user data, e-mail templates, bla bla bla) as you could normally do with JHipster, but you don't have to reinvent the wheel when it comes to common security scenarios, where the JWT option is just not giving you the tools. OAuth2 is used here, because that specification does describe how to handle that.
> 
> I'm still the opinion, that UAA gives a great value for JHipster microservice. It's not exactly the same as OAuth2 option.
> 
> And while I'm here... I'm planning to set up the UAA blueprint first as it was before it was removed. I am aware that Spring is developing a new version of the authorization server. They wanted to remove it because they thought Keycloak does the job, but as I explained so much in the past: there ARE cases when users want to implement their own OAuth2 authserver and not using a complete external solution. I'm happy to see, that the spring team finally got convinced by the com that this still is a thing and proceed their work on a new service. But as version 0.0.1 doesn't seem stable to me I think to restore what there was, and when things get stable, move to that new stuff then.

@xetys @AlexandreCassagne Any advice and help from you will be appreciated too! 

Even though Spring Authorization Server is a new player and 0.0.1 may not be as stable as the latest JHipster UAA server I'm leaning more towards bringing Spring Authorization Server as a blueprint and get it added to the JHipster ecosystem because then it's a win-win for both the parties (at least in my opinion) because with JHipster comes more adoption and more adoption puts more momentum into the spring project. Also, with the addition of it into the JHipster ecosystem, UAA experts like you yourself kinda will start contributing to the spring project, eventually being able to match the stability and functionality of UAA. When the spring team sees adoption and all this activity maybe they'll take it more seriously and do a really good job of maintaining the project. This is vaguely my line of thought.

**Comment by gmarziou** [comment]
0.1.1 is really early, the fact that it is a Spring project does not give any guarantee about its future, it's labeled as experimental and could stop any time. 

About adoption by JHipster community, do you think many users would deploy a real app in production based on an experimental security component? I doubt it.

**Comment by xetys** [comment]
@kevin-madhu as I stated before, my first milestone is that the uaa blueprint (https://github.com/k-tel/generator-jhipster-uaa) will behave exactly as it was before it has been removed from the generator. There are a lot of fixes and actions done to integrate that feature into the rest of JHipster generated code base and there is absolutely no need to reinvent those things again.

When this is done, it could be an experimental feature to let the user choose the spring authorization server instead of the old one.

As @gmarziou I stated, I doubt many users would like the idea of using experimental stuff in production, while we have a stable solution which is field proven for years.

**Comment by kevin-madhu** [comment]
Okay @xetys. Got it.

**Comment by niniss** [comment]
Hello, I agree that a standalone UAA blueprint it's the best solution until Spring AUTH server is only on experimental step and can be removed at any time !

**Comment by dc-tonglec** [comment]
> Azure

@jdubois I saw you mentioned about Azure Active Directory. Is there any example/documentation on how to configure Jhipster with Azure AD B2C? Thanks!

**Comment by jdubois** [comment]
@dc-tonglec I did a blog post about this at https://dev.to/azure/using-spring-security-with-azure-active-directory-mga - be sure to use a recent Spring Boot version, I coded some specific support for Azure that will make AAD B2C work (and even improved it last week: https://github.com/spring-projects/spring-boot/pull/27819 )

**Comment by mraible** [comment]
@jdubois You might want to update your tutorial (or write a new one) that uses auth code flow instead of implicit flow. Implicit flow is deprecated and is no longer recommended. https://developer.okta.com/blog/2019/05/01/is-the-oauth-implicit-flow-dead

**Comment by jdubois** [comment]
Thank you @mraible ! Yes, that would be another tutorial.

**Comment by dc-tonglec** [comment]
Thanks a lot @jdubois & @mraible !

I am really looking forward to the new tutorial. ☺

**Comment by miguelk5** [comment]
Sorry for post on 2021 topic, but I still wonder if Jhipster UAA is available anymore ?


![image](https://user-images.githubusercontent.com/25555655/222713988-5eac3731-b41b-452b-9645-2bb8be760a7e.png)


https://www.jhipster.tech/using-uaa/#jhipster-uaa

On this page one of the first thing to do its to create a Jhipster UAA application ... but can't see that option on Eclipse IDE , Intellij plugin or the CLI.

I was trying to follow this tutorial: https://www.baeldung.com/jhipster-uaa-secured-micro-service

**Comment by yosephsa** [comment]
@pascalgrimaud It seems like the Overview Diagram under the Microservices section still references UAA within the JHipster architecture. This diagram may need to be updated in order to avoid confusion https://www.jhipster.tech/microservices-architecture/

**Comment by mraible** [comment]
I remember there being an issue about updating the microservices architecture diagrams. I'm not sure where it is, but I did create a couple of new diagrams as part of [Micro Frontends for Java Microservices](https://auth0.com/blog/micro-frontends-for-java-microservices/).

**Comment by lsorba** [comment]
Sorry for keeping this topic alive.

I am still a fan of the old but good UAA that I have tweaked, thanks for your work @xetys ! 
 @xetys you wrote few years back that you are still using UAA for your projects, did you manage to configure a microservice with Spring Security 6 to auth² with UAA ? A bit in the idea of @jfougere ?

My goal would be to keep the UAA (I know it is not up to date) but make it work with a generated & adpated JHipster 8 microservice using Spring Security 6,...

If anyone achieved this, I would be very happy to get some tips and support!
